src/ParallelTempering.jl

```diff
@@ -0,0 +1,168 @@
+        MPI.Recv!(rex_accept, rex_rank, 2, MPI.COMM_WORLD)
+    else
+        # receive energy from neighbor
+        E_rex = [0.0]
+        MPI.Recv!(E_rex, rex_rank, 1, MPI.COMM_WORLD)
+
+        # replica exch. acceptance probability
+        P_rex = exp( (replica.nn_βs[rex_ID] - replica.sampler.β)*(E_rex[1] - E_curr[1]) )
+
+        # acceptance criterion.
+        if (P_rex > 1.0) || (rand() <= P_rex)
+            rex_accept[1] = true
+        end
+
+        # send accept/reject info to neighbor
+        MPI.Send(rex_accept, rex_rank, 2, MPI.COMM_WORLD)
+    end
+
+    # replica exch. rejected
+    if !rex_accept[1]
+        return false
+    end
+
+    # accept replica exch.: swap configurations
+    MPI.Sendrecv!(replica.sampler.system.sites, rex_rank, 3,
+                  replica.sampler.system.sites, rex_rank, 3, MPI.COMM_WORLD)
+
+    # recalculate energy of new state
+    replica.sampler.E = energy(replica.sampler.system)
+
+    return true
+end
+
+
+"""
+Start a parallel tempering (PT) simulation. 
+Run w/ the command "mpiexec -n [n_procs] julia --project [julia_script.jl]".
+"""
+function run!(replica::Replica, T_sched::Function;
+                therm_mcs=1000, measure_interval=10, rex_interval=100, max_mcs=500_000, bin_size=1.0)
+    # set replica sampling β using temperature (β = 1.0/(kT)) 
+    T = T_sched(replica.rank+1, replica.N_ranks)
+    set_temp!(replica.sampler, T)
+
+    # set replica neighbor β's  
+    for i in 1:2
+        if replica.nn_ranks[i] != -1
+            replica.nn_βs[i] = 1.0 / T_sched(replica.nn_ranks[i]+1, replica.N_ranks)
+        end
+    end
+
+    # equilibrate replica to it's distribution
+    thermalize!(replica.sampler, therm_mcs)
+
+    N_rex = max_mcs / rex_interval
+    N_measure = cld(rex_interval, measure_interval)
+    replica.sampler.nsweeps = measure_interval
+
+    A = BinnedArray{Float64, Int64}(bin_size=bin_size)
+    rex_accepts = zeros(Int64, 2)
+
+    # start PT with finite length
+    for i in 1:N_rex
+
+        # measurements between exchanges
+        for j in 1:N_measure
+            # decorrelation steps
+            sample!(replica.sampler)
+
+            # add measurements here 
+            #...
+
+            # histogram for WHAM (TODO)
+            A[running_energy(replica.sampler)] += 1
+        end
+
+        # attempt replica exchange
+        if replica_exchange!(replica)
+            rex_accepts[replica.rex_dir+1] += 1
+        end
+
+        # alternate up/down pairs of replicas for exchanges
+        replica.rex_dir = 1 - replica.rex_dir
+    end
+
+    # write replica exchange rates to file for each process
+    f = open(@sprintf("P%03d_rex.dat", replica.rank), "w")
+    println(f, "REX accepts (down, up) = (", rex_accepts[1], " ", rex_accepts[2],"), total = ", sum(rex_accepts), " / ", N_rex)
+    close(f)
+
+    # write histogram to file for each process
+    f = open(@sprintf("P%03d_hist.dat", replica.rank), "w")
+    println(f, A)
+    close(f)
```

Comment:
Would it be preferable for all the processes to communicate their measurements back to a master rank, which then writes them all out into a single file?

Reply:
Also: if the analysis of the output data is going to also be done in Julia, it seems like it might be better to use `Serialization` to `serialize` out the `BinnedArray` to make it easy to re-load (and probably uses less storage).

Though, maybe users want to do the analysis with something else, in which case plain-text makes sense. Serialized files also don't transfer between different versions of Julia / probably different versions of packages. Maybe this is a flag to set? Not sure.

Reply:
Maybe we should have a longer discussion on Zoom with Matt about various design choices? I guess it can get improved in a follow-on pull request.

Reply:
The serialization sounds like a good idea. I'm open to changing the IO however makes the most sense, but had them all writing their own files at first just for my convenience. I'm open to discuss this or other ideas too